core/drivers/clk/clk.c

```diff
@@ -0,0 +1,615 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Bootlin
+ */
+
+#include <drivers/clk.h>
+#include <kernel/boot.h>
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <stddef.h>
+
+#include <drivers/scmi.h>
+
+#define SCMI_MAX_CLK_NAME_LEN	16
+
+static SLIST_HEAD(, clk) clk_list = SLIST_HEAD_INITIALIZER(clk_list);
+
+struct clk *clk_alloc(const char *name, const struct clk_ops *ops,
+		      struct clk **parent_clks, uint8_t parent_count)
+{
+	struct clk *clk;
+	unsigned int parent;
+	unsigned int alloc_size = sizeof(struct clk) +
+				  parent_count * sizeof(struct clk *);
+
+	clk = malloc(alloc_size);
+	if (!clk)
+		panic();
+
+	memset(clk, 0, alloc_size);
+
+	if (parent_count) {
+		clk->parents = (struct clk **)(clk + 1);
+		clk->num_parents = parent_count;
+		for (parent = 0; parent < parent_count; parent++)
+			clk->parents[parent] = parent_clks[parent];
+	}
+
+	clk->name = name;
+	clk->ops = ops;
+	clk->scmi_id = CLK_NO_SCMI_ID;
+
+	return clk;
+}
+
+void clk_free(struct clk *clk)
+{
+	free(clk);
+}
+
+static int clk_check(struct clk *clk)
+{
+	if (clk->ops->set_parent && !clk->ops->get_parent)
+		return 1;
+
+	if (clk->num_parents > 1 && !clk->ops->get_parent)
+		return 1;
+
+	if ((clk->flags & CLK_SET_RATE_PARENT) && !clk->num_parents)
+		return 1;
+
+	return 0;
+}
+
+static void clk_compute_rate(struct clk *clk)
+{
+	unsigned long parent_rate = 0;
+
+	clk->rate = 0;
+
+	if (clk->parent)
+		parent_rate = clk->parent->rate;
+
+	if (clk->ops->get_rate)
+		clk->rate = clk->ops->get_rate(clk, parent_rate);
+	else
+		clk->rate = parent_rate;
+}
+
+uint8_t clk_get_num_parents(struct clk *clk)
+{
+	return clk->num_parents;
+}
+
+struct clk *clk_get_parent_by_index(struct clk *clk, unsigned int pidx)
+{
+	if (pidx >= clk->num_parents)
+		return NULL;
+
+	return clk->parents[pidx];
+}
+
+static void clk_init_parent(struct clk *clk)
+{
+	uint8_t pidx;
+
+	if (clk->num_parents > 1) {
+		pidx = clk->ops->get_parent(clk);
+		if (pidx >= clk->num_parents) {
+			EMSG("get_parent returned an invalid value\n");
+			return;
+		}
+
+		clk->parent = clk->parents[pidx];
+	} else {
+		if (clk->num_parents)
+			clk->parent = clk->parents[0];
+	}
+}
+
+int clk_register(struct clk *clk)
```

Comment:
Use `TEE_Result`  type for error code inside OP-TEE core.

Reply:
OK